pydantic/_internal/_model_construction.py

```diff
@@ -266,6 +266,17 @@ def init_private_attributes(self: BaseModel, __context: Any) -> None:
     object_setattr(self, '__pydantic_private__', pydantic_private)
 
 
+def get_model_post_init(namespace: dict[str, Any], bases: tuple[type[Any], ...]) -> Callable[..., Any] | None:
+    """Get the `model_post_init` method from the namespace or the class bases, or `None` if not defined."""
+    if 'model_post_init' in namespace:
+        return namespace['model_post_init']
+
+    for base in bases:
+        model_post_init = getattr(base, 'model_post_init', None)
+        if model_post_init is not None:
+            return model_post_init
```

Comment:
```suggestion
    model_post_init = get_attribute_from_bases(bases, 'model_post_init')
    if model_post_init is not BaseModel.model_post_init:
        return model_post_init
```
I think this is probably the "most correct" version of this, since we are already using the `get_attribute_from_bases` function which handles mro the same as python, etc.

Reply:
hmm I see this errors, I'll see if I can achieve the same thing in a better way..

Reply:
The above is also closely related to what we do when setting `__pydantic_post_init__` (i.e., ignoring `BaseModel.model_post_init` and follows the same logic as we do when setting `__hash__` (i.e., using the mro).

Reply:
Unfortunately even using `get_attribute_from_bases`, the test is still failing (in fact `get_attribute_from_bases` is doing something quite similar, i.e. using `hasattr` and `getattr`, so we face the same issue). Maybe doing something like the following:

```python
for base in mro_for_bases(bases):
    model_post_init = base.__dict__.get('model_post_init')
    if model_post_init is not None and model_post_init is not BaseModel.model_post_init:
        return model_post_init
```

Reply:
@Viicos this is a good catch — but actually I think this is a bug in the `get_attribute_from_bases` function — it shouldn't be using `hasattr` and `getattr` for exactly this reason. I think looking in the class `__dict__` should fix this and allow us to use that function, if it does, I would propose we make that change. Any objections? If not, I can push a commit and merge this.

Reply:
Sure, do you want to do it in a separate PR? Might be easier if it breaks some tests. Then I'll rebase on `main` once merged

Reply:
Doesn't break any tests — I think my preference would be to just put it in this PR since the test you've added is actually a good test of that change.

To reduce the back-and-forth I'll just push the commit in hopes you are okay with it; I can revert that if you prefer though, and move it elsewhere.

Reply:
Totally ok with it, thanks for the last push!